include/ruby/thread.h

```diff
@@ -264,6 +264,46 @@ rb_internal_thread_event_hook_t *rb_internal_thread_add_event_hook(
 bool rb_internal_thread_remove_event_hook(
         rb_internal_thread_event_hook_t * hook);
 
+
+typedef int rb_internal_thread_specific_key_t;
+#define RB_INTERNAL_THREAD_SPECIFIC_KEY_MAX 8
+/**
+ * Create a key to store thread specific data.
+ *
+ * These APIs are designed for tools using
+ * rb_internal_thread_event_hook APIs.
+ *
+ * Note that only `RB_INTERNAL_THREAD_SPECIFIC_KEY_MAX` keys
+ * can be created. raises `ThreadError` if exceeded.
+ *
+ * Usage:
+ *   // at initialize timing:
+ *   int tool_key; // gvar
+ *   Init_tool() {
+ *     tool_key = rb_internal_thread_specific_key_create();
+ *   }
+ *
+ *   // at thread start timing (with GVL):
+ *   rb_internal_thread_pecific_set(thread, tool_key, per_thread_data);
+ *
+ *   // at any timing:
+ *   per_thread_data = rb_internal_thread_specific_get(thread, tool_key);
+ */
+rb_internal_thread_specific_key_t rb_internal_thread_specific_key_create(void);
+
+/**
+ * Get thread and tool specific data.
+ *
+ * This function is async signal safe and thread safe.
+ */
+void *rb_internal_thread_specific_get(VALUE thread_val, rb_internal_thread_specific_key_t key);
+
+/**
+ * Set thread and tool specific data.
+ * This function should be called with GVL.
```

Comment:
Hum, actually, this may be tricky for `gvltools`, ideally I'd want to be able to initialize the key on the first event received for that thread.

Could we use atomics to allow this?

Reply:
or should we allocate storage for all existing threads (and created threads in future) at `rb_internal_thread_specific_key_create()` is created?

Reply:
This design introduce some more complexity because of Ractors.

Reply:
The current comment is wrong.
Calling this set function is safe if the storage is allocated. For example, if `get()` returns non-null, it is safe to call `set()`.

Reply:
I think preallocating storage would make sense.

Reply:
I don't want to introduce pre-allocation because of complexity. Could you elaborate why it is needed more?

Reply:
Right, so if you want to do record something on `RUBY_INTERNAL_THREAD_EVENT_STARTED` event, you can't because you don't have the GVL, you need to wait for the thread to acquire the GVL first, there you can initialize memory.

(pseudo code)
```c
static void
gt_thread_callback(rb_event_flag_t event, const rb_internal_thread_event_data_t *event_data, void *user_data) {
   gvl_stats stats = rb_internal_thread_specific_get(instrumentation_key);
   if (!stats) {
     stats = alloc_thread_storage();
     rb_internal_thread_specific_set(instrumentation_key, stats);
   }

    switch(event) {
        case RUBY_INTERNAL_THREAD_EVENT_STARTED:
          stats.started_at = time_now();
// snip...
}

Reply:
I see... I missed that STARTED state does not have GVL.

Reply:
Yes. But even beyond that. It is common to have the hook only registered at some point, so the first event of any thread can be any of the events.

Reply:
1. It is easy to pre-allocate at creation timing. No complex issue.
2. It is hard to allocate for *existing* threads.

Does dropping 2 make sense?

Reply:
I think 2 can be worked around by iterating on `Thread.list`.

But couldn't `rb_internal_thread_specific_key_create` lock the whole VM? This is really intended as a debug API so I don't think it's a problem to temporarily stop other threads and ractors. Also there is a max of 8 keys, so it's not like it would happen repeatedly.

Reply:
I added to the limitation "the first call of `rb_internal_thread_specific_key_create()` should not be called with multiple ractors". I think it would be reasonable limitation (Init_foo of extension should call it).